### ISSUE TITLE
File is deleted from storage even though form is not valid

### ISSUE THREAD

**Issue by ste93cry** [dustin10/VichUploaderBundle#878]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 1.8.2

#### Summary

I've created a form which maps directly to a Doctrine entity object using the `data_class` option. The entity properties have some constraints that are checked by the validator when submitting the form to verify that everything is fine before persisting it to the database. In this case I'm considering the update phase, so the entity is not new. The form contains a `VichImageType` field. When submitting the form with some invalid fields and the "delete" checkbox checked, the image is deleted anyway from the storage. This causes troubles for example if the underlying storage throws exceptions when the file can not be found anymore but its filename is still set in the entity being loaded (e.g. Flysystem) and anyway no persistent changes should be done until the form passes the validation

#### Current behavior

The file is deleted from the storage even though the form is not valid (e.g. some constraints fail)

#### How to reproduce

Just create a form with a `VichImageType` or `VichFileType` field that allows deleting the file and some other fields with validation constraints attached, then load it with an existing file binded to the field and submit it by making sure that some of the fields generate an error to prevent the validation step from passing and that the delete file checkbox is checked

#### Expected behavior

The file should be deleted from the storage only if the form does not contain any error

#### Proposed solution

At first I though that the fix was as simple as checking that the form was valid before deleting the file from the storage in the `POST_SUBMIT` event listener added by the `VichFileType` type. However at that point the form has not still been processed completely and so other fields may not have been validated yet. I've managed to workaround the issue by creating a `FormType` [form extension](https://symfony.com/doc/current/form/create_form_type_extension.html) which adds (nearly) the same `POST_SUBMIT` listener added in the `VichFileType` class. The difference is that the listener iterates all form children starting from the root, checks if they are a `VichFileType` field and calls the `$uploadHandler->remove()` method only if the whole form is valid. I've done this because there is no way (at least easy) to add a listener to the root form from a child form when building it. I can prepare a PR if you're fine with this solution

**Comment by garak** [comment]
It looks like a duplicate of #756

**Comment by ste93cry** [comment]
I was sure I saw something like this a few time ago but could not find the PR/issue, I will look better next time, sorry. Should I close this issue and move the discussion about my proposal there?
